patches/tModLoader/Terraria/ModLoader/IO/TileIO.cs

```diff
@@ -12,106 +12,261 @@ namespace Terraria.ModLoader.IO
+
+			unloadedTileIDs = new int[5] {
+					ModContent.Find<ModTile>("ModLoader/UnloadedNonSolidTile").Type,
+					ModContent.Find<ModTile>("ModLoader/UnloadedSemiSolidTile").Type,
+					ModContent.Find<ModTile>("ModLoader/UnloadedChest").Type,
+					ModContent.Find<ModTile>("ModLoader/UnloadedDresser").Type,
+					ModContent.Find<ModTile>("ModLoader/UnloadedTile").Type
+			};
+
+			//NOTE: Refactor the insides of two ForEach into one method if adding another
+
+			// Retrieve Basic Tile Type Data from saved Tile Map, and store in table
 			foreach (var tileTag in tag.GetList<TagCompound>("tileMap")) {
-				ushort type = (ushort)tileTag.GetShort("value");
+				tableData tabData = tableData.Create();
+
+				int saveType = tileTag.GetShort("value") - TileID.Count;
 				string modName = tileTag.GetString("mod");
 				string name = tileTag.GetString("name");
-				tables.tiles[type] = ModContent.TryFind(modName, name, out ModTile tile) ? tile.Type : (ushort)0;
-				if (tables.tiles[type] == 0) {
-					tables.tiles[type] = ModContent.Find<ModTile>("ModLoader/PendingUnloadedTile").Type;
-					tables.tileModNames[type] = modName;
-					tables.tileNames[type] = name;
+				
+				tabData.id = ModContent.TryFind(modName, name, out ModTile tile) ? tile.Type : (ushort)0;
+
+				// Handle unloaded tiles at the tile level by storing restoration data
+				if (tabData.id == 0) { 
+					tabData.modName = modName;
+					tabData.name = name;
+
+					string unloadedType = "ModLoader/UnloadedTile"; 
+					ushort fallbackType = 0;
+
+					if (tileTag.ContainsKey("unloadedType")) { // If statement is for legacy support
+						unloadedType = tileTag.Get<string>("unloadedType");
+						
+						fallbackType = tileTag.Get<ushort>("fallbackType");
+					}
+
+					UpdateUnloaded pendingInfo = new UpdateUnloaded(tileInfos, TilesContext);
+					pendingInfo.AddInfos(new UnloadedInfo(modName, name, fallbackType));
+
+					tabData.fallbackID = fallbackType;
+					tabData.id = ModContent.Find<ModTile>(unloadedType).Type;
 				}
-				tables.frameImportant[type] = tileTag.GetBool("framed");
+
+				tables.tileData[saveType] = tabData;
+				tables.frameImportant[saveType] = tileTag.GetBool("framed");
 			}
+
+			// Retrieve Basic Wall Type Data from saved Wall Map, and store in table
 			foreach (var wallTag in tag.GetList<TagCompound>("wallMap")) {
-				ushort type = (ushort)wallTag.GetShort("value");
+				tableData tabData = tableData.Create();
+
+				int saveType = wallTag.GetShort("value") - WallID.Count;
 				string modName = wallTag.GetString("mod");
 				string name = wallTag.GetString("name");
-				tables.walls[type] = ModContent.TryFind(modName, name, out ModWall wall) ? wall.Type : (ushort)0;
+								
+				tabData.id = ModContent.TryFind(modName, name, out ModWall wall) ? wall.Type : (ushort)0;
+				
+				if (tabData.id == 0) {
+					tabData.id = ModContent.Find<ModWall>("ModLoader/UnloadedWall").Type;
+					tabData.modName = modName;
+					tabData.name = name;
```

Comment:
should `tabData` have an `unloadedInfo` field instead?

Reply:
Not applicable anymore.